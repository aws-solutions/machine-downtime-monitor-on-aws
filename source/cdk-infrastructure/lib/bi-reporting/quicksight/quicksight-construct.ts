// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, CustomResource } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';

/**
 * BI reporting - QuickSight reporting construct props interface
 */
export interface IQuickSightReportingProps {
  // Glue custom resource
  readonly glueCustomResource: CustomResource;
  // Glue database
  readonly glueDatabaseName: string;
  // Glue table
  readonly glueTableName: string;
  // Metadata configuration
  readonly metadataConfiguration: {
    bucketName: string;
    manifestPrefix: string;
    machineInformationPrefix: string;
    machineConfigInformationPrefix: string;
  };
  // QuickSight principal ARN
  readonly quickSightPrincipalArn: string;
  // QuickSight template
  readonly quickSightTemplate: string;
  // Solution helper Lambda function
  readonly solutionHelperFunction: LambdaFunction;
}

/**
 * BI reporting - QuickSight reporting construct
 */
export class QuickSightReporting extends Construct {
  constructor(scope: Construct, id: string, props: IQuickSightReportingProps) {
    super(scope, id);

    const { bucketName, manifestPrefix, machineInformationPrefix, machineConfigInformationPrefix } = props.metadataConfiguration;

    const solutionHelperQuickSightPolicy = new Policy(this, 'SolutionHelperQuickSightPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'quicksight:CreateDataSource',
            'quicksight:DeleteDataSource',
            'quicksight:PassDataSource'
          ],
          resources: [`arn:${Aws.PARTITION}:quicksight:${Aws.REGION}:${Aws.ACCOUNT_ID}:datasource/${Aws.STACK_NAME}*`]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'quicksight:CreateDataSet',
            'quicksight:DeleteDataSet',
            'quicksight:PassDataSet'
          ],
          resources: [`arn:${Aws.PARTITION}:quicksight:${Aws.REGION}:${Aws.ACCOUNT_ID}:dataset/${Aws.STACK_NAME}*`]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'quicksight:CreateAnalysis',
            'quicksight:DeleteAnalysis'
          ],
          resources: [`arn:${Aws.PARTITION}:quicksight:${Aws.REGION}:${Aws.ACCOUNT_ID}:analysis/${Aws.STACK_NAME}*`]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'quicksight:CreateDashboard',
            'quicksight:DeleteDashboard'
          ],
          resources: [`arn:${Aws.PARTITION}:quicksight:${Aws.REGION}:${Aws.ACCOUNT_ID}:dashboard/${Aws.STACK_NAME}*`]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['quicksight:DescribeTemplate'],
          resources: [props.quickSightTemplate]
        })
      ],
      roles: [props.solutionHelperFunction.role!]
    });

    const quickSightCustomResource = new CustomResource(this, 'QuickSightCustomResource', {
      serviceToken: props.solutionHelperFunction.functionArn,
      properties: {
        Action: 'CREATE_QUICKSIGHT',
        AccountId: Aws.ACCOUNT_ID,
        GlueDatabaseName: props.glueDatabaseName,
        GlueTableName: props.glueTableName,
        Metadata: {
          BucketName: bucketName,
          MachineInformationPrefix: machineInformationPrefix,
          MachineConfigInformationPrefix: machineConfigInformationPrefix,
          ManifestPrefix: manifestPrefix
        },
        PrincipalArn: props.quickSightPrincipalArn,
        QuickSightTemplate: props.quickSightTemplate,
        StackName: Aws.STACK_NAME
      }
    });
    quickSightCustomResource.node.addDependency(solutionHelperQuickSightPolicy);
    quickSightCustomResource.node.addDependency(props.glueCustomResource);
  }
}