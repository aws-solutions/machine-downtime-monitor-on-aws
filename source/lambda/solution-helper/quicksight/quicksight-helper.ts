// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import QuickSight from 'aws-sdk/clients/quicksight';
import { IQuickSightCreateAnalysisProps, IQuickSightCreateDashboardProps, IQuickSightCreateDataSetProps,
  IQuickSightCreateDataSourceProps, IQuickSightHelperProps, QuickSightResourceType } from './quicksight-properties';
import { getOptions } from '../../util/metrics';

/**
 * QuickSightHelper class
 * @class QuickSightHelper
 * @description This class handles QuickSight resources.
 */
export default class QuickSightHelper {
  private awsAccountId: string;
  private quickSightClient: QuickSight;
  private quickSightPrincipalArn: string;

  /**
   * QuickSightHelper class constructor
   * @param {IQuickSightHelperProps} props the class properties
   */
  constructor(props: IQuickSightHelperProps) {
    if (!props.awsAccountId || !props.quickSightPrincipalArn) {
      throw new Error('AWS account ID or QuickSight principal ARN cannot be empty.');
    }

    this.awsAccountId = props.awsAccountId;
    this.quickSightClient = new QuickSight(getOptions());
    this.quickSightPrincipalArn = props.quickSightPrincipalArn;
  }

  /**
   * Creates a QuickSight data source.
   * @param {IQuickSightCreateDataSourceProps} props QuickSight data source creation properties
   */
  async createDataSource(props: IQuickSightCreateDataSourceProps) {
    const params: QuickSight.CreateDataSourceRequest = {
      AwsAccountId: this.awsAccountId,
      DataSourceId: props.name,
      Name: props.name,
      Type: props.type,
      DataSourceParameters: props.dataSourceParameters,
      Permissions: this.getPermissions(QuickSightResourceType.DATA_SOURCE),
      SslProperties: { DisableSsl: false }
    };

    console.log('Creating QuickSight data source', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.createDataSource(params).promise();
    console.log('QuickSight data source created', JSON.stringify(response, null, 2));

    return response;
  }

  /**
   * Deletes a QuickSight data source.
   * @param {string} dataSourceId QuickSight data source ID
   */
  async deleteDataSource(dataSourceId: string) {
    const params: QuickSight.DeleteDataSourceRequest = {
      AwsAccountId: this.awsAccountId,
      DataSourceId: dataSourceId
    };

    console.log('Deleting QuickSight data source', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.deleteDataSource(params).promise();
    console.log('QuickSight data source deleted', JSON.stringify(response, null, 2));
  }

  /**
   * Creates a QuickSight data set.
   * @param {IQuickSightCreateDataSetProps} props QuickSight data set creation properties
   */
  async createDataSet(props: IQuickSightCreateDataSetProps) {
    const params: QuickSight.CreateDataSetRequest = {
      AwsAccountId: this.awsAccountId,
      DataSetId: props.name,
      Name: props.name,
      ImportMode: props.importMode,
      PhysicalTableMap: props.physicalTableMap,
      Permissions: this.getPermissions(QuickSightResourceType.DATA_SET),
      LogicalTableMap: props.logicalTableMap
    };

    console.log('Creating QuickSight data set', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.createDataSet(params).promise();
    console.log('QuickSight data set created', JSON.stringify(response, null, 2));

    return response;
  }

  /**
   * Deletes a QuickSight data set.
   * @param {string} dataSetId QuickSight data set ID
   */
  async deleteDataSet(dataSetId: string) {
    const params: QuickSight.DeleteDataSetRequest = {
      AwsAccountId: this.awsAccountId,
      DataSetId: dataSetId
    };

    console.log('Deleting QuickSight data set', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.deleteDataSet(params).promise();
    console.log('QuickSight data set deleted', JSON.stringify(response, null, 2));
  }

  /**
   * Creates a QuickSight analysis.
   * @param {IQuickSightCreateAnalysisProps} props QuickSight analysis creation properties
   */
  async createAnalysis(props: IQuickSightCreateAnalysisProps) {
    const params: QuickSight.CreateAnalysisRequest = {
      AwsAccountId: this.awsAccountId,
      AnalysisId: props.name,
      Name: props.name,
      SourceEntity: props.sourceEntity,
      Permissions: this.getPermissions(QuickSightResourceType.ANALYSIS)
    };

    console.log('Creating QuickSight analysis', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.createAnalysis(params).promise();
    console.log('QuickSight analysis created', JSON.stringify(response, null, 2));

    return response;
  }

  /**
   * Deletes a QuickSight analysis.
   * @param {string} analysisId QuickSight analysis ID
   */
  async deleteAnalysis(analysisId: string) {
    const params: QuickSight.DeleteAnalysisRequest = {
      AwsAccountId: this.awsAccountId,
      AnalysisId: analysisId,
      ForceDeleteWithoutRecovery: true
    };

    console.log('Deleting QuickSight analysis', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.deleteAnalysis(params).promise();
    console.log('QuickSight analysis deleted', JSON.stringify(response, null, 2));
  }

  /**
   * Creates a QuickSight dashboard.
   * @param {IQuickSightCreateDashboardProps} props QuickSight dashboard creation properties
   */
  async createDashboard(props: IQuickSightCreateDashboardProps) {
    const params: QuickSight.CreateDashboardRequest = {
      AwsAccountId: this.awsAccountId,
      DashboardId: props.name,
      Name: props.name,
      SourceEntity: props.sourceEntity,
      Permissions: this.getPermissions(QuickSightResourceType.DASHBOARD),
      DashboardPublishOptions: {
        AdHocFilteringOption: { AvailabilityStatus: 'ENABLED' },
        ExportToCSVOption: { AvailabilityStatus: 'ENABLED' },
        SheetControlsOption: { VisibilityState: 'EXPANDED' }
      }
    };

    console.log('Creating QuickSight dashboard', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.createDashboard(params).promise();
    console.log('QuickSight dashboard created', JSON.stringify(response, null, 2));

    return response;
  }

  /**
   * Deletes a QuickSight dashbaord.
   * @param {string} dashboardId QuickSight dashbaord ID
   */
  async deleteDashboard(dashboardId: string) {
    const params: QuickSight.DeleteDashboardRequest = {
      AwsAccountId: this.awsAccountId,
      DashboardId: dashboardId
    };

    console.log('Deleting QuickSight dashbaord', JSON.stringify(params, null, 2));
    const response = await this.quickSightClient.deleteDashboard(params).promise();
    console.log('QuickSight dashbaord deleted', JSON.stringify(response, null, 2));
  }

  /**
   * Gets QuickSight permissions based on the resource type.
   * @param {QuickSightResourceType} resourceType QuickSight resource type
   * @returns {QuickSight.ResourcePermissionList} QuickSight resource permissions
   */
  getPermissions(resourceType: QuickSightResourceType): QuickSight.ResourcePermissionList {
    switch(resourceType) {
      case QuickSightResourceType.DATA_SOURCE:
        return [{
          Principal: this.quickSightPrincipalArn,
          Actions: [
            'quicksight:DescribeDataSource',
            'quicksight:DescribeDataSourcePermissions',
            'quicksight:PassDataSource',
            'quicksight:UpdateDataSource',
            'quicksight:DeleteDataSource',
            'quicksight:UpdateDataSourcePermissions'
          ]
        }];
      case QuickSightResourceType.DATA_SET:
        return [{
          Principal: this.quickSightPrincipalArn,
          Actions: [
            'quicksight:DescribeDataSet',
            'quicksight:DescribeDataSetPermissions',
            'quicksight:PassDataSet',
            'quicksight:DescribeIngestion',
            'quicksight:ListIngestions',
            'quicksight:UpdateDataSet',
            'quicksight:DeleteDataSet',
            'quicksight:CreateIngestion',
            'quicksight:CancelIngestion',
            'quicksight:UpdateDataSetPermissions'
          ]
        }];
      case QuickSightResourceType.ANALYSIS:
        return [{
          Principal: this.quickSightPrincipalArn,
          Actions: [
            'quicksight:RestoreAnalysis',
            'quicksight:UpdateAnalysisPermissions',
            'quicksight:DeleteAnalysis',
            'quicksight:QueryAnalysis',
            'quicksight:DescribeAnalysisPermissions',
            'quicksight:DescribeAnalysis',
            'quicksight:UpdateAnalysis'
          ]
        }];
      case QuickSightResourceType.DASHBOARD:
        return [{
          Principal: this.quickSightPrincipalArn,
          Actions: [
            'quicksight:DescribeDashboard',
            'quicksight:ListDashboardVersions',
            'quicksight:UpdateDashboardPermissions',
            'quicksight:QueryDashboard',
            'quicksight:UpdateDashboard',
            'quicksight:DeleteDashboard',
            'quicksight:DescribeDashboardPermissions',
            'quicksight:UpdateDashboardPublishedVersion'
          ]
        }];
      default:
        throw new Error(`Unsupported resource type: ${resourceType}`);
    }
  }
}