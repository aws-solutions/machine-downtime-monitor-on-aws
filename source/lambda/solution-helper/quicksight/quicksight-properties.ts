// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import QuickSight from 'aws-sdk/clients/quicksight';

/**
 * @interface IQuickSightHelperProps
 * @description QuickSightHelper properties
 */
export interface IQuickSightHelperProps {
  /**
   * AWS Account ID
   */
  awsAccountId: string;
  /**
   * QuickSight principal ARN for permissions
   */
  quickSightPrincipalArn: string;
}

/**
 * @interface IQuickSightCreateDataSourceProps
 * @extends IQuickSightResourceCommonProps
 * @description The properties to create a QuickSight data source
 */
export interface IQuickSightCreateDataSourceProps extends IQuickSightResourceCommonProps {
  /**
   * QuickSight data source type
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_CreateDataSource.html#QS-CreateDataSource-request-Type
   */
  type: QuickSightDataSourceType;
  /**
   * QuickSight data source parameters
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_DataSourceParameters.html
   */
  dataSourceParameters: QuickSight.DataSourceParameters;
}

/**
 * @interface IQuickSightCreateDataSetProps
 * @extends IQuickSightResourceCommonProps
 * @description The properties to create a QuickSight data set
 */
export interface IQuickSightCreateDataSetProps extends IQuickSightResourceCommonProps {
  /**
   * QuickSight data set import mode
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_CreateDataSet.html#QS-CreateDataSet-request-ImportMode
   */
  importMode: QuickSightDataSetImportMode;
  /**
   * QuickSight data set physical table map
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_PhysicalTable.html
   */
  physicalTableMap: QuickSight.PhysicalTableMap;
  /**
   * (Optional) QuickSight data set column groups
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_ColumnGroup.html
   */
  columnGroups?: QuickSight.ColumnGroupList;
  /**
   * (Optional) QuickSight data set logical table map
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_LogicalTable.html
   */
  logicalTableMap?: QuickSight.LogicalTableMap;
}

/**
 * @interface IQuickSightCreateAnalysisProps
 * @extends IQuickSightResourceCommonProps
 * @description The properties to create a QuickSight analysis
 */
export interface IQuickSightCreateAnalysisProps extends IQuickSightResourceCommonProps {
  /**
   * QuickSight analysis source entity
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_AnalysisSourceEntity.html
   */
  sourceEntity: QuickSight.AnalysisSourceEntity;
}

/**
 * @interface IQuickSightCreateDashboardProps
 * @extends IQuickSightResourceCommonProps
 * @description The properties to create a QuickSight dashboard
 */
export interface IQuickSightCreateDashboardProps extends IQuickSightResourceCommonProps {
  /**
   * QuickSight dashboard source entity
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_DashboardSourceEntity.html
   */
  sourceEntity: QuickSight.DashboardSourceEntity;
  /**
   * (Optional) QuickSight dashboard publish options
   * https://docs.aws.amazon.com/quicksight/latest/APIReference/API_DashboardPublishOptions.html
   */
  dashboardPublishOptions?: QuickSight.DashboardPublishOptions;
}

/**
 * @interface IQuickSightResourceCommon
 * @description The common properties of QuickSight resource
 */
interface IQuickSightResourceCommonProps {
  name: string;
}

/**
 * @enum
 * @description The QuickSight resource type
 */
export enum QuickSightResourceType {
  DATA_SOURCE = 'DATA_SOURCE',
  DATA_SET = 'DATA_SET',
  ANALYSIS = 'ANALYSIS',
  DASHBOARD = 'DASHBOARD'
}

/**
 * @enum
 * @description The QuickSight data source type
 */
export enum QuickSightDataSourceType {
  ADOBE_ANALYTICS = 'ADOBE_ANALYTICS',
  AMAZON_ELASTICSEARCH = 'AMAZON_ELASTICSEARCH',
  ATHENA = 'ATHENA',
  AURORA = 'AURORA',
  AURORA_POSTGRESQL = 'AURORA_POSTGRESQL',
  AWS_IOT_ANALYTICS = 'AWS_IOT_ANALYTICS',
  GITHUB = 'GITHUB',
  JIRA = 'JIRA',
  MARIADB = 'MARIADB',
  MYSQL = 'MYSQL',
  ORACLE = 'ORACLE',
  POSTGRESQL = 'POSTGRESQL',
  PRESTO = 'PRESTO',
  REDSHIFT = 'REDSHIFT',
  S3 = 'S3',
  SALESFORCE = 'SALESFORCE',
  SERVICENOW = 'SERVICENOW',
  SNOWFLAKE = 'SNOWFLAKE',
  SPARK = 'SPARK',
  SQLSERVER = 'SQLSERVER',
  TERADATA = 'TERADATA',
  TWITTER = 'TWITTER',
  TIMESTREAM = 'TIMESTREAM'
}

/**
 * @enum
 * @description The QuickSight data set import mode
 */
export enum QuickSightDataSetImportMode {
  SPICE = 'SPICE',
  DIRECT_QUERY = 'DIRECT_QUERY'
}