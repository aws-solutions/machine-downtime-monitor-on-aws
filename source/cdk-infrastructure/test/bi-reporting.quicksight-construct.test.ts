// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { CustomResource, Stack } from '@aws-cdk/core';
import { Code, Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';
import { QuickSightReporting } from '../lib/bi-reporting/quicksight/quicksight-construct';

test('BI reporting QuickSight snapshot and variables', () => {
  // PREPARE
  const stack = new Stack();
  const glueCustomResource = new CustomResource(stack, 'TestCustomResource', {
    serviceToken: 'test-service-token'
  });
  const sourceCodeBucket = Bucket.fromBucketName(stack, 'TestSourceBucket', 'test-bucket');
  const solutionHelperFunction = new LambdaFunction(stack, 'TestSolutionHelperFunction', {
    code: Code.fromBucket(sourceCodeBucket, 'source.zip'),
    runtime: Runtime.NODEJS_14_X,
    handler: 'index.handler'
  });

  // WHEN
  new QuickSightReporting(stack, 'TestQuickSight', {
    glueCustomResource: glueCustomResource,
    glueDatabaseName: 'test-database',
    glueTableName: 'test-table',
    metadataConfiguration: {
      bucketName: 'test-bucket',
      machineInformationPrefix: 'machine_information',
      machineConfigInformationPrefix: 'machine_config_information',
      manifestPrefix: 'manifest'
    },
    quickSightPrincipalArn: 'test-quicksight-principal-arn',
    quickSightTemplate: 'test-quicksight-template',
    solutionHelperFunction: solutionHelperFunction,
  });

  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});