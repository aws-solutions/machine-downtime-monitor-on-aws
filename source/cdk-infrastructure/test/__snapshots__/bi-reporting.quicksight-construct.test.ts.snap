// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BI reporting QuickSight snapshot and variables 1`] = `
Object {
  "Resources": Object {
    "TestCustomResource": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": "test-service-token",
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestQuickSightQuickSightCustomResourceA152450C": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "TestCustomResource",
        "TestQuickSightSolutionHelperQuickSightPolicyB3914B77",
      ],
      "Properties": Object {
        "AccountId": Object {
          "Ref": "AWS::AccountId",
        },
        "Action": "CREATE_QUICKSIGHT",
        "GlueDatabaseName": "test-database",
        "GlueTableName": "test-table",
        "Metadata": Object {
          "BucketName": "test-bucket",
          "MachineConfigInformationPrefix": "machine_config_information",
          "MachineInformationPrefix": "machine_information",
          "ManifestPrefix": "manifest",
        },
        "PrincipalArn": "test-quicksight-principal-arn",
        "QuickSightTemplate": "test-quicksight-template",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestSolutionHelperFunction4CFAB303",
            "Arn",
          ],
        },
        "StackName": Object {
          "Ref": "AWS::StackName",
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestQuickSightSolutionHelperQuickSightPolicyB3914B77": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "quicksight:CreateDataSource",
                "quicksight:DeleteDataSource",
                "quicksight:PassDataSource",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":quicksight:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":datasource/",
                    Object {
                      "Ref": "AWS::StackName",
                    },
                    "*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "quicksight:CreateDataSet",
                "quicksight:DeleteDataSet",
                "quicksight:PassDataSet",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":quicksight:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":dataset/",
                    Object {
                      "Ref": "AWS::StackName",
                    },
                    "*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "quicksight:CreateAnalysis",
                "quicksight:DeleteAnalysis",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":quicksight:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":analysis/",
                    Object {
                      "Ref": "AWS::StackName",
                    },
                    "*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "quicksight:CreateDashboard",
                "quicksight:DeleteDashboard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":quicksight:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":dashboard/",
                    Object {
                      "Ref": "AWS::StackName",
                    },
                    "*",
                  ],
                ],
              },
            },
            Object {
              "Action": "quicksight:DescribeTemplate",
              "Effect": "Allow",
              "Resource": "test-quicksight-template",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestQuickSightSolutionHelperQuickSightPolicyB3914B77",
        "Roles": Array [
          Object {
            "Ref": "TestSolutionHelperFunctionServiceRole7B84FD28",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestSolutionHelperFunction4CFAB303": Object {
      "DependsOn": Array [
        "TestSolutionHelperFunctionServiceRole7B84FD28",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-bucket",
          "S3Key": "source.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestSolutionHelperFunctionServiceRole7B84FD28",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestSolutionHelperFunctionServiceRole7B84FD28": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
